### PR TITLE
handle platform dependencies + update feature graph construction

### LINKED ISSUE #64: Add model for new feature resolver once it lands upstream
At the moment we have a pretty good model for Cargo's current feature resolver, including the feature unification that happens across normal/dev/build dependencies within a workspace.

This task tracks the new feature resolver in https://github.com/rust-lang/cargo/pull/7820. We'll probably want to model this as well once it lands upstream.

### PR DESCRIPTION
**Pull request by sunshowers**
This is a complete overhaul of the way platform-specific dependencies are
handled. Based on my experimentation and reading the Cargo source code, I
believe that this is now correct.

Doing so also required feature graph construction to be updated, so the
feature graph construction is ready for the new feature resolver, and is now
platform-sensitive as well. The APIs for the feature graph are still to
come, but making the data model be full-fidelity allows for both the current
and new feature resolvers to be implemented.

Now that target logic is internalized in guppy, cargo-guppy no longer needs
to do its own evaluation.

For more about the new feature resolver, see:

* https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#features
* https://github.com/rust-lang/cargo/issues/7914
* https://github.com/rust-lang/cargo/issues/7915
* https://github.com/rust-lang/cargo/issues/7916

Closes #64.

### CODE REVIEW

**Comment by sunshowers** [comment]
I'd want to get https://github.com/EmbarkStudios/cfg-expr/pull/6 landed + include the three-value logic support in https://github.com/sunshowers/cfg-expr/tree/logic.

File: guppy/src/graph/graph_impl.rs [hunk 1]
```diff
@@ -809,27 +818,325 @@ impl DependencyMetadata {
         &self.version_req
     }
 
-    /// Returns true if this is an optional dependency.
+    /// Returns true if this is an optional dependency on the platform `guppy` is running on.
+    ///
+    /// This will also return true if this dependency will never be included on this platform at
+    /// all. To get finer-grained information, use the `enabled` method instead.
     pub fn optional(&self) -> bool {
```
Comment (metajack) [inline guppy/src/graph/graph_impl.rs]: Does there need to be an optional_on()?

File: target-spec/src/parser.rs [hunk 2]
```diff
@@ -112,7 +112,7 @@ fn expression(input: &str) -> IResult<&str, Expr> {
     alt((any, all, not, test_equal, test_set))(input)
 }
 
-fn spec(input: &str) -> IResult<&str, Target> {
+fn spec(input: &str) -> IResult<&str, TargetEnum> {
```
[the remark this answers is not in the document]
Comment (sunshowers) [reply target-spec/src/parser.rs]: I'll just call it `Target` I think.

File: guppy/src/graph/graph_impl.rs (hunk 1 above)
Comment (sunshowers) [reply guppy/src/graph/graph_impl.rs:825]: For symmetry reasons, I guess. Will add.

Comment (sunshowers) [reply guppy/src/graph/graph_impl.rs:826]: Done!

File: target-spec/src/parser.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (sunshowers) [reply target-spec/src/parser.rs]: Done.

**Comment by sunshowers** [comment]
Thanks for the review! Before landing, I did some further refactoring so that `TargetSpec::eval` no longer returns errors -- all error checking is now done upfront. Will land shortly, thanks again!

Just need to wait on EmbarkStudios/cfg-expr#7 to land and be published to crates.io now. At that point we can release a guppy update ^_^

**Pull request merged**